src/test/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandlerTests.java

```diff
@@ -0,0 +1,143 @@
+import org.springframework.messaging.Message;
+
+/**
+ * @author Gary Russell
+ * @since 2.0
+ *
+ */
+public class KafkaProducerMessageHandlerTests {
+
+	private static String topic1 = "testTopic1out";
+
+	private static String topic2 = "testTopic2out";
+
+	@ClassRule
+	public static KafkaEmbedded embeddedKafka = new KafkaEmbedded(1, true, topic1, topic2);
+
+	private static Consumer<Integer, String> consumer;
+
+	private static CountDownLatch consumerLatch = new CountDownLatch(1);
+
+	@BeforeClass
+	public static void setUp() throws Exception {
+		ConsumerFactory<Integer, String> cf = new DefaultKafkaConsumerFactory<>(
+				KafkaTestUtils.consumerProps("testOut", "true", embeddedKafka));
+		consumer = cf.createConsumer();
+		consumer.subscribe(Arrays.asList(topic1, topic2), new ConsumerRebalanceListener() {
+
+			@Override
+			public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+			}
+
+			@Override
+			public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+				consumerLatch.countDown();
+			}
+
+		});
+		consumer.poll(0); // force assignment
+		assertThat(consumerLatch.await(30, TimeUnit.SECONDS)).isTrue();
+	}
+
+	@Test
+	public void testOutbound() {
+		ProducerFactory<Integer, String> producerFactory = new DefaultKafkaProducerFactory<>(
+				KafkaTestUtils.producerProps(embeddedKafka));
+		KafkaTemplate<Integer, String> template = new KafkaTemplate<>(producerFactory);
+		KafkaProducerMessageHandler<Integer, String> handler = new KafkaProducerMessageHandler<>(template);
+		handler.setBeanFactory(mock(BeanFactory.class));
+		handler.afterPropertiesSet();
+
+		Message<?> message = MessageBuilder.withPayload("foo")
+				.setHeader(KafkaHeaders.TOPIC, topic1)
+				.setHeader(KafkaHeaders.MESSAGE_KEY, 2)
+				.setHeader(KafkaHeaders.PARTITION_ID, 1)
+				.build();
+		handler.handleMessage(message);
+
+		ConsumerRecord<Integer, String> record = getSingleRecord();
+		assertThat(record).has(key(2));
+		assertThat(record).has(partition(1));
+		assertThat(record).has(value("foo"));
+
+		message = MessageBuilder.withPayload("bar")
+				.setHeader(KafkaHeaders.TOPIC, topic1)
+				.setHeader(KafkaHeaders.PARTITION_ID, 0)
+				.build();
+		handler.handleMessage(message);
+		record = getSingleRecord();
+		assertThat(record).has(key((Integer) null));
+		assertThat(record).has(partition(0));
+		assertThat(record).has(value("bar"));
+
+		message = MessageBuilder.withPayload("baz")
+				.setHeader(KafkaHeaders.TOPIC, topic1)
+				.build();
+		handler.handleMessage(message);
+		record = getSingleRecord();
+		assertThat(record).has(key((Integer) null));
+		assertThat(record).has(value("baz"));
+	}
+
+	private ConsumerRecord<Integer, String> getSingleRecord() {
+		ConsumerRecords<Integer, String> received = getRecords();
+		assertThat(received.count()).isEqualTo(1);
+		ConsumerRecord<Integer, String> record = received.records(topic1).iterator().next();
+		return record;
```

Comment:
Not sure if STS can be configured do not extract variable automatically on the method invocation, but my IDEA always barks to me that such a construction should be simplified.